File: server/pipelines/rampart_annotate.py
#!/usr/bin/env python3
"""
RAMPART Annotation Script
Standalone script to map FASTQ reads to references using mappy and produce RAMPART-compatible CSV annotations.
"""

import argparse
import gzip
import sys
from collections import defaultdict
from pathlib import Path

try:
    import mappy
except ImportError:
    print("Error: mappy is not installed. Install it with: pip install mappy", file=sys.stderr)
    sys.exit(1)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Map FASTQ reads to references and create RAMPART annotation CSV'
    )
    parser.add_argument(
        '-i', '--input',
        required=True,
        help='Input FASTQ or FASTQ.gz file'
    )
    parser.add_argument(
        '-r', '--reference',
        required=True,
        help='Reference FASTA file'
    )
    parser.add_argument(
        '-o', '--output',
        required=True,
        help='Output CSV file'
    )
    parser.add_argument(
        '-b', '--barcode',
        default='unknown',
        help='Barcode name (default: unknown)'
    )
    parser.add_argument(
        '-m', '--min-identity',
        type=float,
        default=0.0,
        help='Minimum mapping identity (0.0-1.0, default: 0.0)'
    )
    parser.add_argument(
        '-t', '--threads',
        type=int,
        default=2,
        help='Number of threads (default: 2)'
    )
    parser.add_argument(
        '--reference-fields',
        default='display_name',
        help='Reference fields to extract from FASTA headers (format: field_name[header_key], default: display_name[display_name])'
    )

    # Mappy / minimap2 sensitivity options
    parser.add_argument(
        '-k', '--kmer',
        type=int,
        default=None,
        help='Minimizer k-mer length (default: preset value, typically 15 for map-ont). '
             'Lower values (e.g. 13) increase sensitivity for divergent sequences.'
    )
    parser.add_argument(
        '-w', '--window',
        type=int,
        default=None,
        help='Minimizer window size (default: preset value). '
             'Lower values increase sensitivity at the cost of speed.'
    )
    parser.add_argument(
        '--min-cnt',
        type=int,
        default=None,
        help='Minimum number of minimizers on a chain (default: preset value, typically 3). '
             'Lower values (e.g. 2) increase sensitivity.'
    )
    parser.add_argument(
        '--min-chain-score',
        type=int,
        default=None,
        help='Minimum chaining score (default: preset value). '
             'Lower values allow shorter/weaker alignments through.'
    )
    parser.add_argument(
        '--min-dp-score',
        type=int,
        default=None,
        help='Minimum dynamic programming alignment score (default: preset value). '
             'Lower values allow weaker DP alignments.'
    )
    parser.add_argument(
        '--best-n',
        type=int,
        default=1,
        help='Report at most this many alignments per read (default: 1, i.e. best hit only).'
    )
    parser.add_argument(
        '--score-n',
        type=int,
        default=1,
        help='Penalty for aligning against ambiguous bases (N). '
             'Set to 0 (--score-n 0) to ignore N mismatches, which is useful for '
             'viral reference genomes with N-padded regions. '
             'Corresponds to the 7th element of the minimap2 scoring vector (default: 1).'
    )
    return parser.parse_args()


def open_fastq(filepath):
    """Open FASTQ or FASTQ.gz file."""
    filepath = Path(filepath)
    if filepath.suffix == '.gz':
        return gzip.open(filepath, 'rt')
    return open(filepath, 'r')


def parse_fastq_simple(file_handle):
    """Simple FASTQ parser that yields (name, sequence, quality, header)."""
    while True:
        header = file_handle.readline().strip()
        if not header:
            break
        seq = file_handle.readline().strip()
        plus = file_handle.readline().strip()
        qual = file_handle.readline().strip()
        
        # Parse read name (first field before space)
        read_name = header[1:].split()[0]
        yield read_name, seq, qual, header[1:]


def parse_reference_fasta_headers(reference_file):
    """Parse reference FASTA headers to extract metadata."""
    ref_info = defaultdict(dict)
    ref_lengths = {}
    
    with open(reference_file, 'r') as f:
        current_id = None
        for line in f:
            line = line.strip()
            if line.startswith('>'):
                # Parse header
                header = line[1:]
                tokens = header.split()
                current_id = tokens[0]  # First token is the sequence ID
                
                # Parse key=value pairs
                for token in tokens[1:]:
                    if '=' in token:
                        key, value = token.split('=', 1)
                        ref_info[current_id][key] = value
                
                # If no display_name, use the ID
                if 'display_name' not in ref_info[current_id]:
                    ref_info[current_id]['display_name'] = current_id
                
                ref_lengths[current_id] = 0
            elif current_id:
                ref_lengths[current_id] += len(line)
    
    return ref_info, ref_lengths


def parse_reference_fields_spec(spec):
    """Parse reference fields specification.
    Format: field1[key1],field2[key2];field3[key3]
    Returns: dict of {output_field: input_key}
    """
    if not spec:
        return {}
    
    fields = {}
    for item in spec.replace(';', ',').split(','):
        item = item.strip()
        if '[' in item and ']' in item:
            field_name = item[:item.index('[')]
            key_name = item[item.index('[')+1:item.index(']')]
            fields[field_name] = key_name
        else:
            # Default: use same name for both
            fields[item] = item
    
    return fields


def calculate_identity_from_cigar(cs_tag):
    """Calculate identity from minimap2 cs tag.
    Format: cs:Z::10*ag:5+tc:3-cc:2
    """
    if not cs_tag:
        return 0, 0
    
    # Remove cs:Z: prefix
    if cs_tag.startswith('cs:Z:'):
        cs_tag = cs_tag[5:]
    
    matches = 0
    mismatches = 0
    
    i = 0
    while i < len(cs_tag):
        if cs_tag[i] == ':':
            # Match run
            i += 1
            num_str = ''
            while i < len(cs_tag) and cs_tag[i].isdigit():
                num_str += cs_tag[i]
                i += 1
            if num_str:
                matches += int(num_str)
        elif cs_tag[i] == '*':
            # Substitution
            mismatches += 1
            i += 3  # Skip *XY
        elif cs_tag[i] == '+':
            # Insertion (no penalty for identity calc)
            i += 1
            while i < len(cs_tag) and cs_tag[i].islower():
                i += 1
        elif cs_tag[i] == '-':
            # Deletion (no penalty for identity calc)
            i += 1
            while i < len(cs_tag) and cs_tag[i].islower():
                i += 1
        else:
            i += 1
    
    if matches + mismatches == 0:
        return 0, 0
    
    identity = matches / (matches + mismatches)
    return matches, identity


def annotate_reads(args):
    """Main annotation function."""
    
    # Load reference information
    print(f"Loading reference: {args.reference}", file=sys.stderr)
    ref_info, ref_lengths = parse_reference_fasta_headers(args.reference)
    
    # Parse reference fields specification
    ref_fields_spec = parse_reference_fields_spec(args.reference_fields)
    
    # Initialize minimap2 aligner with mappy
    print(f"Initializing minimap2 with {args.threads} threads...", file=sys.stderr)

    # Build aligner kwargs — only pass values that were explicitly set
    aligner_kwargs = {
        'preset': 'map-ont',
        'n_threads': args.threads,
        'best_n': args.best_n,
    }

    if args.kmer is not None:
        aligner_kwargs['k'] = args.kmer
    if args.window is not None:
        aligner_kwargs['w'] = args.window
    if args.min_cnt is not None:
        aligner_kwargs['min_cnt'] = args.min_cnt
    if args.min_chain_score is not None:
        aligner_kwargs['min_chain_score'] = args.min_chain_score
    if args.min_dp_score is not None:
        aligner_kwargs['min_dp_score'] = args.min_dp_score

    # Build the scoring vector if N-penalty differs from default (1), or if any
    # explicit scoring option requires it.  minimap2 default scoring for map-ont is
    # approximately [2, 4, 4, 2, 24, 1, 1]. We only construct the vector when
    # score_n != 1 to avoid overriding the preset's other scoring defaults.
    if args.score_n != 1:
        # [match, mismatch, gap_open, gap_extend, long_gap_open, long_gap_extend, N_penalty]
        aligner_kwargs['scoring'] = [2, 4, 4, 2, 24, 1, args.score_n]
        print(f"  N-mismatch penalty set to {args.score_n} (scoring vector: {aligner_kwargs['scoring']})",
              file=sys.stderr)

    aligner = mappy.Aligner(args.reference, **aligner_kwargs)
    
    if not aligner:
        print(f"Error: Failed to load reference {args.reference}", file=sys.stderr)
        sys.exit(1)
    
    # Create output directory if it doesn't exist
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    print(f"Created output directory: {output_path.parent}", file=sys.stderr)
    
    # Open input and output files
    print(f"Processing: {args.input}", file=sys.stderr)
    
    counts = {'total': 0, 'mapped': 0, 'unmapped': 0}
    
    with open_fastq(args.input) as fq, open(args.output, 'w') as csv_out:
        # Write CSV header
        header_fields = ['read_name', 'read_len', 'start_time', 'barcode', 
                        'best_reference', 'ref_len', 'start_coords', 'end_coords', 
                        'num_matches', 'mapping_len']
        
        # Add reference fields to header
        for field_name in ref_fields_spec.keys():
            header_fields.append(field_name)
        
        csv_out.write(','.join(header_fields) + '\n')
        
        # Process each read
        for read_name, seq, qual, full_header in parse_fastq_simple(fq):
            counts['total'] += 1
            
            read_len = len(seq)
            
            # Extract start_time from header if present
            start_time = '?'
            if 'start_time=' in full_header:
                for token in full_header.split():
                    if token.startswith('start_time='):
                        start_time = token.split('=', 1)[1]
                        break
            
            # Map the read (cs=True requests the cs tag for identity calculation)
            alignments = list(aligner.map(seq, cs=True))
            
            # Handle mappings
            if not alignments:
                # No mapping
                counts['unmapped'] += 1
                row = [
                    read_name, str(read_len), start_time, args.barcode,
                    '?', str(max(ref_lengths.values()) if ref_lengths else 0),
                    '0', '0', '0', '0'
                ]
                # Add '?' for reference fields
                for _ in ref_fields_spec:
                    row.append('?')
                csv_out.write(','.join(row) + '\n')
                
            elif len(alignments) > 1:
                # Multiple primary mappings - ambiguous
                row = [
                    read_name, str(read_len), start_time, args.barcode,
                    '?', str(alignments[0].ctg_len),
                    '0', '0', '0', '0'
                ]
                for _ in ref_fields_spec:
                    row.append('?')
                csv_out.write(','.join(row) + '\n')
                
            else:
                # Single best mapping
                aln = alignments[0]
                
                # mappy provides cs tag via the cs attribute if available
                cs_tag = None
                if hasattr(aln, 'cs') and aln.cs:
                    cs_tag = f"cs:Z:{aln.cs}"
                
                if cs_tag:
                    num_matches, identity = calculate_identity_from_cigar(cs_tag)
                else:
                    # Fallback: use match count from alignment
                    # aln.mlen is the number of matching bases
                    # aln.blen is the alignment block length
                    num_matches = aln.mlen if hasattr(aln, 'mlen') else 0
                    if hasattr(aln, 'blen') and aln.blen > 0:
                        identity = aln.mlen / aln.blen
                    else:
                        # Use NM (edit distance) if available
                        mapping_len = aln.r_en - aln.r_st
                        if hasattr(aln, 'NM'):
                            identity = 1.0 - (aln.NM / mapping_len) if mapping_len > 0 else 0
                        else:
                            identity = aln.mapq / 60.0 if hasattr(aln, 'mapq') else 0
                        num_matches = int(mapping_len * identity)
                
                # Check identity threshold
                if identity < args.min_identity:
                    counts['unmapped'] += 1
                    row = [
                        read_name, str(read_len), start_time, args.barcode,
                        '*', '0', '0', '0', '0', '0'
                    ]
                    for _ in ref_fields_spec:
                        row.append('*')
                    csv_out.write(','.join(row) + '\n')
                    continue
                
                counts['mapped'] += 1
                
                # Calculate mapping length
                mapping_len = aln.r_en - aln.r_st
                
                row = [
                    read_name,
                    str(read_len),
                    start_time,
                    args.barcode,
                    aln.ctg,
                    str(aln.ctg_len),
                    str(aln.r_st),
                    str(aln.r_en),
                    str(num_matches),
                    str(mapping_len)
                ]
                
                # Add reference fields
                for field_name, key_name in ref_fields_spec.items():
                    value = ref_info.get(aln.ctg, {}).get(key_name, '?')
                    row.append(value)
                
                csv_out.write(','.join(row) + '\n')
            
            if counts['total'] % 1000 == 0:
                print(f"Processed {counts['total']} reads...", file=sys.stderr)
    
    # Print summary
    print(f"\nSummary:", file=sys.stderr)
    print(f"  Total reads: {counts['total']}", file=sys.stderr)
    print(f"  Mapped: {counts['mapped']} ({100*counts['mapped']/counts['total']:.1f}%)", file=sys.stderr)
    print(f"  Unmapped: {counts['unmapped']} ({100*counts['unmapped']/counts['total']:.1f}%)", file=sys.stderr)
    print(f"\nOutput written to: {args.output}", file=sys.stderr)


if __name__ == '__main__':
    args = parse_args()
    annotate_reads(args)

File: server/pipelines/rampart_annotate.js
#!/usr/bin/env node
/**
 * RAMPART Annotation Script (Node.js)
 *
 * Maps FASTQ reads to references using minimap2 (spawned directly) and produces
 * RAMPART-compatible CSV annotations. Replaces the mappy-based Python version so
 * no Python installation is required on any platform.
 *
 * CLI usage (mirrors the Python version):
 *   node rampart_annotate.js -i reads.fastq -r references.fasta -o out.csv -b barcode01
 *
 * Can also be required as a module:
 *   const { run } = require('./rampart_annotate');
 *   await run({ inputFile, referenceFile, outputFile, ... });
 */

'use strict';

const { spawn }  = require('child_process');
const fs         = require('fs');
const path       = require('path');
const zlib       = require('zlib');
const readline   = require('readline');

// ─── FASTA reference parser ───────────────────────────────────────────────────

/**
 * Read a FASTA file and extract per-sequence metadata.
 * Header format expected:  >seqId key1=value1 key2=value2 ...
 * Returns { refInfo: {seqId: {key: val}}, refLengths: {seqId: number} }
 */
function parseReferenceFasta(refPath) {
    const refInfo    = {};
    const refLengths = {};
    let   currentId  = null;

    const content = fs.readFileSync(refPath, 'utf8');
    for (const rawLine of content.split('\n')) {
        const line = rawLine.trim();
        if (!line) continue;

        if (line.startsWith('>')) {
            const tokens  = line.slice(1).split(/\s+/);
            currentId     = tokens[0];
            refInfo[currentId]    = {};
            refLengths[currentId] = 0;

            for (const token of tokens.slice(1)) {
                const eq = token.indexOf('=');
                if (eq !== -1) {
                    refInfo[currentId][token.slice(0, eq)] = token.slice(eq + 1);
                }
            }
            // Fallback: display_name defaults to sequence id
            if (!refInfo[currentId].display_name) {
                refInfo[currentId].display_name = currentId;
            }
        } else if (currentId) {
            refLengths[currentId] += line.length;
        }
    }

    return { refInfo, refLengths };
}

// ─── FASTQ header scanner ─────────────────────────────────────────────────────

/**
 * Stream through a FASTQ or FASTQ.gz and build a Map<readName, startTime>.
 * Only the header lines (every 4th line starting at 0) are examined.
 */
function buildStartTimeIndex(fastqPath) {
    return new Promise((resolve, reject) => {
        const index = new Map();
        const raw   = fs.createReadStream(fastqPath);
        const src   = fastqPath.endsWith('.gz') ? raw.pipe(zlib.createGunzip()) : raw;
        const rl    = readline.createInterface({ input: src, crlfDelay: Infinity });

        let lineNum = 0;
        rl.on('line', (line) => {
            if (lineNum % 4 === 0 && line.startsWith('@')) {
                const readName = line.slice(1).split(/\s/)[0];
                const m = line.match(/start_time=(\S+)/);
                index.set(readName, m ? m[1] : '?');
            }
            lineNum++;
        });

        rl.on('close', () => resolve(index));
        rl.on('error', reject);
        src.on('error', reject);
        raw.on('error', reject);
    });
}

// ─── Reference-fields spec parser ────────────────────────────────────────────

/**
 * Parse a reference-fields spec like "display_name[display_name],taxon[taxon_id]"
 * Returns [{ outputField, inputKey }, ...]
 */
function parseRefFieldsSpec(spec) {
    if (!spec || !spec.trim()) {
        return [{ outputField: 'display_name', inputKey: 'display_name' }];
    }
    const fields = [];
    for (const item of spec.replace(/;/g, ',').split(',')) {
        const t = item.trim();
        if (!t) continue;
        const bOpen = t.indexOf('[');
        if (bOpen !== -1 && t.endsWith(']')) {
            fields.push({ outputField: t.slice(0, bOpen), inputKey: t.slice(bOpen + 1, -1) });
        } else {
            fields.push({ outputField: t, inputKey: t });
        }
    }
    return fields.length ? fields : [{ outputField: 'display_name', inputKey: 'display_name' }];
}

// ─── Main run function ────────────────────────────────────────────────────────

/**
 * Run the annotation pipeline.
 *
 * @param {object}   opts
 * @param {string}   opts.inputFile      FASTQ or FASTQ.gz path
 * @param {string}   opts.referenceFile  FASTA reference path
 * @param {string}   opts.outputFile     Output CSV path
 * @param {string}   [opts.barcode]      Barcode label (default: 'unknown')
 * @param {number}   [opts.threads]      minimap2 threads (default: 2)
 * @param {string}   [opts.minimap2Path] Path to minimap2 binary (default: 'minimap2')
 * @param {string}   [opts.refFieldsSpec] Reference fields spec (default: 'display_name[display_name]')
 * @param {number}   [opts.minIdentity]  Minimum alignment identity 0–1 (default: 0)
 * @param {function} [opts.onMessage]    Progress callback (string) -> void
 * @returns {Promise<{total, mapped, unmapped}>}
 */
async function run({
    inputFile,
    referenceFile,
    outputFile,
    barcode      = 'unknown',
    threads      = 2,
    minimap2Path = 'minimap2',
    refFieldsSpec = 'display_name[display_name]',
    minIdentity  = 0.0,
    onMessage    = () => {},
}) {
    // 1. Parse reference FASTA for metadata
    onMessage('Loading reference FASTA...');
    const { refInfo, refLengths } = parseReferenceFasta(referenceFile);
    const maxRefLen = Object.values(refLengths).reduce((a, b) => Math.max(a, b), 0);
    const refFields = parseRefFieldsSpec(refFieldsSpec);

    // 2. Scan FASTQ headers to build readName → startTime index
    onMessage('Scanning FASTQ headers for timestamps...');
    const startTimeIndex = await buildStartTimeIndex(inputFile);

    // 3. Prepare output CSV
    fs.mkdirSync(path.dirname(outputFile), { recursive: true });
    const csvHeader = [
        'read_name', 'read_len', 'start_time', 'barcode',
        'best_reference', 'ref_len', 'start_coords', 'end_coords',
        'num_matches', 'mapping_len',
        ...refFields.map(f => f.outputField),
    ].join(',');

    const outStream = fs.createWriteStream(outputFile);
    outStream.write(csvHeader + '\n');

    // 4. Run minimap2 in PAF mode
    //    --paf-no-hit  → one output line per query even if unmapped
    //    --cs          → include cs tag (for future identity from cigars)
    //    -N 1          → report at most 1 (best) alignment per read
    const mm2Args = [
        '-x', 'map-ont',
        '--cs',
        '--paf-no-hit',
        '-N', '1',
        '-t', String(threads),
        referenceFile,
        inputFile,
    ];

    onMessage(`Mapping with minimap2 (${threads} threads)...`);

    const counts = { total: 0, mapped: 0, unmapped: 0 };

    await new Promise((resolve, reject) => {
        // On Windows, the bundled minimap2.exe co-locates its MinGW/MSYS2 runtime DLLs
        // next to the binary. We must add that directory to PATH so Windows finds them.
        const spawnEnv = { ...process.env };
        const mm2Dir = path.dirname(minimap2Path);
        if (mm2Dir && mm2Dir !== '.') {
            spawnEnv.PATH = mm2Dir + path.delimiter + (spawnEnv.PATH || '');
        }
        const mm2 = spawn(minimap2Path, mm2Args, { env: spawnEnv });

        let pafBuf = '';

        const processLine = (line) => {
            if (!line.trim()) return;
            const cols = line.split('\t');
            if (cols.length < 12) return;

            const readName     = cols[0];
            const readLen      = cols[1];
            const targetName   = cols[5];   // '*' when unmapped
            const targetLen    = cols[6];
            const targetStart  = cols[7];
            const targetEnd    = cols[8];
            const numMatches   = cols[9];
            const alignBlockLen = parseInt(cols[10], 10) || 0;

            const startTime  = startTimeIndex.get(readName) || '?';
            const mappingLen = parseInt(targetEnd, 10) - parseInt(targetStart, 10);

            counts.total++;

            if (targetName === '*') {
                // Unmapped read
                counts.unmapped++;
                const row = [
                    readName, readLen, startTime, barcode,
                    '?', String(maxRefLen), '0', '0', '0', '0',
                    ...refFields.map(() => '?'),
                ];
                outStream.write(row.join(',') + '\n');
            } else {
                // Check identity threshold
                const identity = alignBlockLen > 0
                    ? (parseInt(numMatches, 10) / alignBlockLen)
                    : 0;

                if (identity < minIdentity) {
                    counts.unmapped++;
                    const row = [
                        readName, readLen, startTime, barcode,
                        '*', '0', '0', '0', '0', '0',
                        ...refFields.map(() => '*'),
                    ];
                    outStream.write(row.join(',') + '\n');
                } else {
                    counts.mapped++;
                    const meta = refInfo[targetName] || {};
                    const row = [
                        readName, readLen, startTime, barcode,
                        targetName, targetLen, targetStart, targetEnd,
                        numMatches, String(mappingLen),
                        ...refFields.map(f => meta[f.inputKey] || '?'),
                    ];
                    outStream.write(row.join(',') + '\n');
                }
            }
        };

        // Buffer stdout and flush complete lines
        mm2.stdout.on('data', (chunk) => {
            pafBuf += chunk.toString();
            const lines = pafBuf.split('\n');
            pafBuf = lines.pop(); // hold the (possibly incomplete) last fragment
            for (const line of lines) processLine(line);
        });

        mm2.stdout.on('end', () => {
            // Flush any remaining complete line
            if (pafBuf.trim()) processLine(pafBuf);
            outStream.end();
        });

        const stderrChunks = [];
        mm2.stderr.on('data', (d) => stderrChunks.push(d.toString()));

        mm2.on('error', (err) => {
            reject(new Error(
                `minimap2 failed to start (path: "${minimap2Path}"): ${err.message}\n` +
                'Check that minimap2 is installed and on PATH, or the bundled binary is intact.'
            ));
        });

        mm2.on('exit', (code) => {
            if (code !== 0) {
                const stderr = stderrChunks.join('').trim();
                reject(new Error(`minimap2 exited with code ${code}. stderr:\n${stderr}`));
                return;
            }
            onMessage(
                `Done: ${counts.total} reads processed ` +
                `(${counts.mapped} mapped, ${counts.unmapped} unmapped/low-identity)`
            );
            resolve();
        });
    });

    return counts;
}

// ─── CLI entry point ──────────────────────────────────────────────────────────

function parseCLIArgs(argv) {
    const args = {};
    let i = 2; // skip 'node' and script path
    while (i < argv.length) {
        const flag = argv[i];
        const val  = argv[i + 1];
        switch (flag) {
            case '-i': case '--input':            args.inputFile      = val; i += 2; break;
            case '-r': case '--reference':        args.referenceFile  = val; i += 2; break;
            case '-o': case '--output':           args.outputFile     = val; i += 2; break;
            case '-b': case '--barcode':          args.barcode        = val; i += 2; break;
            case '-t': case '--threads':          args.threads        = parseInt(val, 10); i += 2; break;
            case '--minimap2':                    args.minimap2Path   = val; i += 2; break;
            case '--reference-fields':            args.refFieldsSpec  = val; i += 2; break;
            case '-m': case '--min-identity':     args.minIdentity    = parseFloat(val); i += 2; break;
            default: i++;
        }
    }
    return args;
}

if (require.main === module) {
    const opts = parseCLIArgs(process.argv);

    if (!opts.inputFile || !opts.referenceFile || !opts.outputFile) {
        console.error('Usage: node rampart_annotate.js -i <fastq> -r <ref.fasta> -o <out.csv> [-b barcode] [-t threads] [--minimap2 path] [--reference-fields spec] [-m min_identity]');
        process.exit(1);
    }

    run({ ...opts, onMessage: (msg) => process.stderr.write(msg + '\n') })
        .then((counts) => {
            process.stderr.write(
                `Summary: ${counts.total} reads, ${counts.mapped} mapped, ${counts.unmapped} unmapped\n`
            );
            process.exit(0);
        })
        .catch((err) => {
            process.stderr.write(`Error: ${err.message}\n`);
            process.exit(1);
        });
}

module.exports = { run };
